[ios][expo-media-library] Introducing pairedVideoAsset

entiendoNull opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Currently, when using `expo-media-library`, we don't have access to the video asset of Live Photos, which would be pretty sweet to have now that `expo-live-photo` was recently released. This PR attempts to bring this functionality to the Media Library as well.

I requested this feature and @tsapeta suggested I could try open a PR for it. Regardless how this goes, I had a lot of fun with this 😄

# How
<!--
How did you build this feature or fix this bug and why?
-->

The implementation adds support for Live Photos by:

- Detecting Live Photos using PHAsset's mediaSubtypes
- Exposing the pairedVideo through a new `pairedVideoAsset`-property in the `AssetInfo`

I found it reasonable to expose the `pairedVideoAsset` through `getAssetInfoAsync` - this new property is added to the response of it. Currently, this property is automatically populated if the image has a media subtype of `livePhoto`. Would it be better to make this opt-in instead?

The `PairedVideoAsset` type is designed to look like the `ImagePickerAsset` type for familiarity.

https://github.com/user-attachments/assets/2812076b-ff7e-49bf-9e41-c960d6657b14

I have not committed the changes to the screen in `native-components-list` (which you see in the video). That was just a test I did.

It should also be said that this is my first PR ever touching Swift. I am not very familiar with it, and not sure about best practices or the conventions that you may have  - so please review it thorough! 😅 Also, any feedback that you might have is much appreciated - I'd really love to learn more.

If it's worth continuing with this PR I should also
- update docs
- add a screen to `native-components-list` describing the functionality.

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

I tested this through Bare-Expo on my phone and it seems to work well. 
However, many of the tests for the Media Library seems to fail when I run them. I will have to look into this.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

entiendoNull committed:
[ios][expo-media-library] Introducing pairedVideoAsset that can be used with e.g. expo-live-photo

entiendoNull committed:
Clean up

entiendoNull committed:
Fix formatting

entiendoNull committed:
Formatting

entiendoNull committed:
Formatting...

entiendoNull committed:
Breaking handleLivePhoto out to different methods

tsapeta commented on packages/expo-media-library/src/MediaLibrary.ts:
```diff
@@ -145,6 +145,63 @@ export type AssetInfo = Asset & {
    */
   orientation?: number;
+  /**
+   * This field is available if the type is an image, and the subtype is livePhoto
+   * Whether the asset is stored on the network (iCloud on iOS).
```
```suggestion
   * Contains information about the video paired with the image file.
   * This field is available if the `mediaType` is `"photo"`, and the `mediaSubtypes` includes `"livePhoto"`.
```

tsapeta commented on packages/expo-media-library/src/MediaLibrary.ts:
```diff
@@ -145,6 +145,63 @@ export type AssetInfo = Asset & {
+ * Represents a video component of a Live Photo on iOS. This is the paired video asset that
+ * accompanies a Live Photo's still image, containing the motion content.
+ */
+export type PairedVideoAsset = {
```
I guess we can remove some properties by merging it with the base type
```suggestion
export type PairedVideoAsset = Asset & {
```

entiendoNull replied:
Yeah, that's a very good point. I was so focused on what it looked like in the ImagePicker that I missed this. This sounds reasonable.

tsapeta commented on packages/expo-media-library/src/MediaLibrary.ts:
```diff
@@ -145,6 +145,63 @@ export type AssetInfo = Asset & {
+   * Constant value 'pairedVideo' indicating this is extracted from a Live Photo
+   * @platform ios
+   */
+  type: 'pairedVideo';
```
Maybe this should be a new subtype instead?

entiendoNull replied:
Yeah, if it'll be an extension of the Asset type I think this makes good sense too. So, a new property called `subtype` (that for now will only be "pairedVideo")?

entiendoNull committed:
Apply suggestions from code review

Co-authored-by: Tomasz Sapeta <tomasz.sapeta@me.com>

entiendoNull commented on packages/expo-media-library/src/MediaLibrary.ts:
```diff
@@ -145,6 +145,63 @@ export type AssetInfo = Asset & {
+  /**
+   * File size of the video, in bytes
+   */
+  fileSize?: number;
+  /**
+   * Length of the video in milliseconds or `null` if the asset is not a video
+   * @platform ios
+   */
+  duration?: number;
+  /**
+   * The MIME type of the selected asset or `null` if could not be determined
+   * @platform ios
+   */
+  mimeType?: string | null;
```
Perhaps I should also just remove the mimeType and fileSize? They're not on the originalAsset or the AssetInfo either.

entiendoNull committed:
Updated types exposed to public

entiendoNull committed:
Fixing linting errors + adjusting pairedVideoAsset props

entiendoNull commented on packages/expo-media-library/ios/MediaLibraryModule.swift:
```diff
@@ -339,6 +339,60 @@ public class MediaLibraryModule: Module, PhotoLibraryObserverHandler {
+      }
+      let avAsset = AVAsset(url: fileUrl)
+      let duration = avAsset.duration.seconds
+      var pairedVideoAsset: [String: Any?] = [
```
`[String: Any?]` - saw how that was how it was done in e.g. `exportAsset`, so used the same idea here as well.

tsapeta reviewed (approved):
Looks good to me, just left two comments 👍 I'm requesting an additional review from @behenate as he worked on live photos support.

Don't forget to undraft the PR when you think it's ready 😄

behenate commented on packages/expo-media-library/src/MediaLibrary.ts:
```diff
@@ -145,6 +145,23 @@ export type AssetInfo = Asset & {
+  /**
+   * Constant value 'pairedVideo' indicating this is extracted from a Live Photo
+   */
+  subType: 'pairedVideo';
```
I think I prefer adding a `pairedVideo` to `MediaTypeValue` and setting the type of the video asset as it. That's what we did in `expo-image-picker`. Then we won't have any need for a subType field.

entiendoNull replied:
Absolutely! I misunderstood the previous feedback about subtype where I interpreted that it should be added as a property. I'll have a look adding it to `MediaTypeValue` instead 😊

behenate commented on packages/expo-media-library/ios/MediaLibraryModule.swift:
```diff
@@ -339,6 +339,60 @@ public class MediaLibraryModule: Module, PhotoLibraryObserverHandler {
+        "subType": "pairedVideo"
+      ]
+      if #available(iOS 16, *) {
+        pairedVideoAsset["width"] = videoResource.pixelWidth
+        pairedVideoAsset["height"] = videoResource.pixelHeight
```
I think that the video asset is lower resolution (1080p I think) than the image component of the live-photo

entiendoNull replied:
The way I understand it is that the pixel dimensions of a PHAsset represent the size of the main asset (the photo) and the videoResource contain specifically dimensions to the video component of the Live Photo - which may vary. However, in some tests I just did, the dimensions for the video resource are returning 0 🤔😅 I'll need to look into this.

behenate commented:
Hi we decided offline that we don't want to add the `subType` field. There is already a `mediaSubtypes` field in the asset. Can't we add a `pairedVideo` into the MediaSubtypes type and add it to the `mediaSubtypes` array for paired videos?

entiendoNull commented:
> Hi we decided offline that we don't want to add the `subType` field. There is already a `mediaSubtypes` field in the asset. Can't we add a `pairedVideo` into the MediaSubtypes type and add it to the `mediaSubtypes` array for paired videos?

Absolutely, I will look into this.
Thanks for your feedback @tsapeta  @behenate!

entiendoNull committed:
Code review changes

entiendoNull committed:
Got rid of commented code

Watchara242540 commented:
ปิดทุกระบบให้ผมที่ นี้คือการพิมพ์สั่ง อาบะนาว

ในวันที่ ศ. 13 ธ.ค. 2024 14:53 น. Expo Bot ***@***.***>
เขียนว่า:

> *Hi there! 👋 I'm a bot whose goal is to ensure your contributions meet
> our guidelines.*
>
> I've found some issues in your pull request that should be addressed
> (click on them for more details) 👇
> *⚠️ Suggestion*: Missing changelog entries
>
>
> Your changes should be noted in the changelog, e.g.:
> - [expo-media-library] Introducing pairedVideoAsset ([#33274](
> https://github.com/expo/expo/pull/33274) by ***@***.***(
> https://github.com/entiendoNull))
> Read Updating Changelogs
> <https://github.com/expo/expo/blob/main/guides/contributing/Updating%20Changelogs.md>
> guide and consider adding an appropriate entry to the following changelogs:
>
>    - packages/expo-media-library/CHANGELOG.md
>    <https://github.com/entiendoNull/expo/blob/feature/media-library-live-photo/packages/expo-media-library/CHANGELOG.md>
>
> *⚠️ Suggestion*: SwiftLint Violations
>
>
> Found 4 violations, 0 serious. See the review comments below for more
> insights.
> ------------------------------
>
> *Generated by ExpoBot 🤖 against 839263e
> <https://github.com/expo/expo/commit/839263ea88cc350dc20fbf9d2e122b8f422d42d6>*
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/expo/expo/pull/33274#issuecomment-2540806752>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BJCBR46OUGSQF3VU2RDUED32FKG6NAVCNFSM6AAAAABSQ4DOVOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNBQHAYDMNZVGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

entiendoNull committed:
swiftlint

entiendoNull commented:
@tsapeta @behenate 

- Removed the `subType`
- Kept `pairedVideoAsset` (same name as in expo-image-picker). It populates if:
	- it's iOS
	- the mediaSubType of a photo contains "livePhoto"
- Removed the width/height check from PHAssetResource and retrieve it from AVAsset instead (found function "readSizeFrom" in ExpoImagePicker/VideoUtils). Copied function over to not have cross references.

```
// Usage
<LivePhotoView
  source={{
    photoUri: media.localUri,
    pairedVideoUri: media.pairedVideoAsset.uri,
  }}
/>


// Image Picker Usage
<LivePhotoView
  source={{
    photoUri: media.uri,
    pairedVideoUri: media.pairedVideoAsset.uri,
  }}
/>
```

In the Media Library implementation using just `uri` does not seem to work.
The `uri` in the Media Libaray is `ph://`, the localUri is `file://`.

Also, two tests in the Test Suite fails (not sure if this is already case since before).
![image](https://github.com/user-attachments/assets/9d82ba89-83f8-4732-884b-d645d6f64b88)

![image](https://github.com/user-attachments/assets/0a2a047a-0964-4ae5-aef0-86b2de8b17a2)

Would be great if you have a chance to test it out before it's merged.

entiendoNull committed:
Changelog

entiendoNull commented:
Hey @behenate, just want to fetch your attention so that this PR doesn't go entirely stale! I think I've implemented the changes you requested a while back. Had a git hickup, but I think it's good to go for your review.

Merry belated Christmas and a Happy New Year to you!

behenate reviewed (approved):
Thanks for the reminder! The tests also sometimes fail on main, so I guess this PR doesn't break anything. 
Happy new year to you too!
